src/core/ext/transport/chaotic_good/server/chaotic_good_server.cc

```diff
@@ -347,38 +349,70 @@ ChaoticGoodServerListener::DataConnectionListener::RequestDataConnection() {
       [connection_id, self = WeakRefAsSubclass<DataConnectionListener>()]() {
         self->ConnectionTimeout(connection_id);
       });
-  pending_connections_.emplace(connection_id,
-                               PendingConnectionInfo{latch, timeout_task});
+  pending_connections_.emplace(
+      connection_id,
+      PendingConnectionInfo{
+          latch, timeout_task,
+          handshake_result_args.GetObjectRef<grpc_auth_context>()});
   return PendingConnection(connection_id,
                            Map(latch->Wait(), [latch](auto x) { return x; }));
 }
 
-ChaoticGoodServerListener::DataConnectionListener::PromiseEndpointLatchPtr
+std::optional<
+    ChaoticGoodServerListener::DataConnectionListener::PendingConnectionInfo>
 ChaoticGoodServerListener::DataConnectionListener::Extract(
     absl::string_view id) {
   MutexLock lock(&mu_);
   auto ex = pending_connections_.extract(id);
-  if (!ex.empty()) {
-    event_engine_->Cancel(ex.mapped().timeout);
-    return std::move(ex.mapped().latch);
+  if (ex.empty()) {
+    return std::nullopt;
   }
-  return nullptr;
+
+  event_engine_->Cancel(ex.mapped().timeout);
+  return ex.mapped();
 }
 
 void ChaoticGoodServerListener::DataConnectionListener::ConnectionTimeout(
     absl::string_view id) {
-  auto latch = Extract(id);
-  if (latch != nullptr) {
-    latch->Set(absl::DeadlineExceededError("Connection timeout"));
+  std::optional<PendingConnectionInfo> pending_connection = Extract(id);
+  if (pending_connection == std::nullopt) {
+    GRPC_TRACE_LOG(chaotic_good, ERROR)
+        << "ConnectionTimeout: connection not found";
+    return;
   }
+  pending_connection->latch->Set(
+      absl::DeadlineExceededError("Connection timeout"));
 }
 
 void ChaoticGoodServerListener::DataConnectionListener::FinishDataConnection(
-    absl::string_view id, PromiseEndpoint endpoint) {
-  auto latch = Extract(id);
-  if (latch != nullptr) {
-    latch->Set(std::move(endpoint));
+    absl::string_view id, PromiseEndpoint endpoint,
+    RefCountedPtr<grpc_auth_context> auth_context) {
+  auto pending_connection = Extract(id);
+  if (pending_connection == std::nullopt) {
+    GRPC_TRACE_LOG(chaotic_good, ERROR) << "FinishDataConnection: connection "
+                                           "not found";
+    return;
+  }
+
+  if (pending_connection->control_endpoint_auth_context == nullptr) {
+    GRPC_TRACE_LOG(chaotic_good, INFO)
+        << "Control endpoint auth context is null, skipping auth context "
+           "comparison";
+  } else {
+    // Compare the auth context with the one from the control endpoint.
+    // If they are different, drop the connection.
+    std::optional<bool> is_same_peer = auth_context->CompareAuthContext(
+        pending_connection->control_endpoint_auth_context.get());
+    if (is_same_peer.has_value() && !is_same_peer.value()) {
```

Comment:
I didn't get a chance to review before merge, but it looks like we are ignoring a potential bad status here.